fvm/environment/programs.go

```diff
@@ -215,45 +211,84 @@ func (programs *Programs) getOrLoadAddressProgram(
 		return nil, fmt.Errorf("get program failed: %w", freezeError)
 	}
 
-	// reading program from cache
-	program, programState, has := programs.txnState.GetProgram(address)
-	if has {
+	loader := newProgramLoader(load, programs.dependencyStack)
+	program, err := programs.txnState.GetOrComputeProgram(
+		programs.txnState,
+		address,
+		loader,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error getting program: %w", err)
+	}
+
+	programs.dependencyStack.addDependencies(program.Dependencies)
+
+	if loader.Called() {
+		programs.cacheMiss()
+	} else {
 		programs.cacheHit()
+	}
 
-		programs.dependencyStack.addDependencies(program.Dependencies)
-		err := programs.txnState.AttachAndCommitNestedTransaction(programState)
-		if err != nil {
-			panic(fmt.Sprintf(
-				"merge error while getting program, panic: %s",
-				err))
-		}
+	return program.Program, nil
+}
 
-		return program.Program, nil
+// programLoader is used to load a program from a location.
+type programLoader struct {
+	loadFunc        func() (*interpreter.Program, error)
+	dependencyStack *dependencyStack
+	called          bool
+}
+
+var _ derived.ValueComputer[common.AddressLocation, *derived.Program] = (*programLoader)(nil)
+
+func newProgramLoader(
+	loadFunc func() (*interpreter.Program, error),
+	dependencyStack *dependencyStack,
+) *programLoader {
+	return &programLoader{
+		loadFunc:        loadFunc,
+		dependencyStack: dependencyStack,
+		called:          false,
 	}
-	programs.cacheMiss()
+}
 
-	interpreterProgram, programState, dependencies, err :=
-		programs.loadWithDependencyTracking(address, load)
+func (loader *programLoader) Compute(
+	txState state.NestedTransaction,
+	address common.AddressLocation,
+) (
+	*derived.Program,
+	error,
+) {
+	if loader.called {
```

Comment:
Should add a sanity check that the Compute is called with an address same as the input address to create the loader